buildcache/build_cache.go

```diff
@@ -0,0 +1,83 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package buildcache
+
+import (
+	"time"
+
+	"github.com/arduino/go-paths-helper"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const lastUsedFileName = ".last-used"
+
+// BuildCache represents a cache of built files (sketches and cores), it's designed
+// to work on directories. Given a directory as "base" it handles direct subdirectories as
+// keys
+type BuildCache struct {
+	baseDir *paths.Path
+}
+
+// GetOrCreate retrieves or creates the cache directory at the given path
+// If the cache already exists the lifetime of the cache is extended.
+func (bc *BuildCache) GetOrCreate(key string) (*paths.Path, error) {
+	keyDir := bc.baseDir.Join(key)
+	if err := keyDir.MkdirAll(); err != nil {
+		return nil, err
+	}
+
+	if err := keyDir.Join(lastUsedFileName).WriteFile([]byte{}); err != nil {
+		return nil, err
+	}
+	return keyDir, nil
+}
+
+// Purge removes all cache directories within baseDir that have expired
+// To know how long ago a directory has been last used
+// it checks into the .last-used file.
+func (bc *BuildCache) Purge(ttl time.Duration) {
+	files, err := bc.baseDir.ReadDir()
+	if err != nil {
+		return
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			removeIfExpired(file, ttl)
+		}
+	}
+}
+
+// New instantiates a build cache
+func New(baseDir *paths.Path) *BuildCache {
+	return &BuildCache{baseDir}
+}
+
+func removeIfExpired(dir *paths.Path, ttl time.Duration) {
+	fileInfo, err := dir.Join().Stat()
```

Comment:
```suggestion
	fileInfo, err := dir.Join(lastUsedFileName).Stat()
```

I wasn't able to find any good reference (only informal stuff like [this](https://stackoverflow.com/a/3620704/7059512)), but experiments show that writing to the existing `.last-used` file alone is not enough to change the modification time of the parent folder.

I didn't find that to be a problem for the sketch cache, since some other file operations in there seemed to bump the folder's modification time on every compilation, but the only action taken on the core cache is writing to `.last-used` so the core cache would always be purged once TTL had passed since the original creation of the cache, regardless of whether the cache had been used more recently than TTL.

## Demo

A compilation is done three times at intervals of approximately 30 s, which is less than the configured TTL of 60 s:

```text
$ arduino-cli version
arduino-cli.exe  Version: git-snapshot Commit: d9d07e49 Date: 2023-02-03T16:20:25Z

$ export ARDUINO_BUILD_CACHE_COMPILATIONS_BEFORE_PURGE=1

$ export ARDUINO_BUILD_CACHE_TTL=60s

$ arduino-cli sketch new /tmp/FooSketch
Sketch created in: C:\Users\per\AppData\Local\Temp\FooSketch

$ rm -rf /tmp/arduino/cores

$ SECONDS=0; \
  for i in {1..3}; do {
    echo "Starting compilation. Elapsed: $SECONDS"
    ./arduino-cli compile -b arduino:avr:uno /tmp/FooSketch > /dev/null
    echo "Finished compilation. Core cache does$([ "$(ls -A /tmp/arduino/cores)" ] || echo " NOT") exist"
    sleep 30
  } done

Starting compilation. Elapsed: 0
Finished compilation. Core cache does exist
Starting compilation. Elapsed: 34
Finished compilation. Core cache does exist
Starting compilation. Elapsed: 68
Finished compilation. Core cache does NOT exist
```

🐛 The core cache was purged on the third compilation even though it had been used only ~34 s previously.

Reply:
Thanks @per1234, should be fixed now